File: run-diagnostics.py
#!/usr/bin/env python3
"""
Скрипт для диагностики проблемы с расходами
Выполняет запросы к Supabase API
"""

import urllib.request
import urllib.parse
import json
from collections import defaultdict

SUPABASE_URL = "https://bkmicyguzlwampuindff.supabase.co"
SUPABASE_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImJrbWljeWd1emx3YW1wdWluZGZmIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjQ2OTUwMjMsImV4cCI6MjA4MDI3MTAyM30.v8BlZ_k8DxdSmh5Ao1da7GHurSshE1cBsMxdfQCp9PQ"

def fetch_data(date_from, date_to):
    """Получить данные из Supabase"""
    url = f"{SUPABASE_URL}/rest/v1/ozon_performance_daily"
    campaign_name_encoded = urllib.parse.quote("Кабель КГ 2*2,5")
    params = f"?campaign_name=eq.{campaign_name_encoded}&stat_date=gte.{date_from}&stat_date=lte.{date_to}&select=stat_date,money_spent,sku,campaign_id"

    headers = {
        'apikey': SUPABASE_KEY,
        'Authorization': f'Bearer {SUPABASE_KEY}'
    }

    req = urllib.request.Request(url + params, headers=headers)

    try:
        with urllib.request.urlopen(req) as response:
            return json.loads(response.read().decode())
    except Exception as e:
        print(f"❌ Ошибка при запросе: {e}")
        return []

def deduplicate_expenses(data):
    """Дедупликация расходов по дням"""
    daily_expenses = {}
    for row in data:
        date = row['stat_date']
        expense = float(row.get('money_spent', 0))
        if date not in daily_expenses:
            daily_expenses[date] = expense
        else:
            daily_expenses[date] = max(daily_expenses[date], expense)
    return daily_expenses

print('🔍 ДИАГНОСТИКА ПРОБЛЕМЫ С РАСХОДАМИ')
print('=====================================')
print('Кампания: Кабель КГ 2*2,5')
print('Период: 1-31 декабря 2025')
print('')

# Запрос 1: За неделю (1-7 декабря)
print('📊 ЗАПРОС 4: За неделю (1-7 декабря) - ожидается ~24 428 ₽')
print('----------------------------------------------------------------')
data_week = fetch_data('2025-12-01', '2025-12-07')

if data_week:
    daily_exp = deduplicate_expenses(data_week)
    total = sum(daily_exp.values())
    print(f'Всего дней: {len(daily_exp)}')
    print(f'Общая сумма: {total:.2f} ₽')
    print(f'Ожидалось: ~24 428 ₽')
    print(f'Разница: {total - 24428:.2f} ₽')
    print(f'Совпадает: {"✅ ДА" if abs(total - 24428) < 100 else "❌ НЕТ"}')
else:
    print('❌ Нет данных')
print('')

# Запрос 2: За месяц (1-31 декабря)
print('📊 ЗАПРОС 5: За месяц (1-31 декабря) - ожидается ~109 130 ₽')
print('----------------------------------------------------------------')
data_month = fetch_data('2025-12-01', '2025-12-31')

if data_month:
    daily_exp = deduplicate_expenses(data_month)
    total = sum(daily_exp.values())
    print(f'Всего дней с данными: {len(daily_exp)} из 31')
    print(f'Общая сумма: {total:.2f} ₽')
    print(f'Ожидалось: ~109 130 ₽')
    print(f'Разница: {total - 109130:.2f} ₽')
    print(f'Совпадает: {"✅ ДА" if abs(total - 109130) < 100 else "❌ НЕТ"}')
    print('')

    if len(daily_exp) < 31:
        print('⚠️ ПРОБЛЕМА: Не все дни декабря есть в базе!')
        print(f'   Отсутствует {31 - len(daily_exp)} дней')
        print('')
        print('📅 Даты с данными:')
        for i, (date, expense) in enumerate(sorted(daily_exp.items()), 1):
            print(f'   {i:2}. {date}: {expense:>10.2f} ₽')
        print('')

        # Найдем отсутствующие даты
        all_dates = [f'2025-12-{day:02d}' for day in range(1, 32)]
        missing_dates = [d for d in all_dates if d not in daily_exp]
        if missing_dates:
            print('❌ Отсутствующие даты:')
            for date in missing_dates:
                print(f'   - {date}')
            print('')
    else:
        print('✅ Все 31 день декабря присутствуют в базе')
        print('')
else:
    print('❌ Нет данных')
print('')

# Запрос 3: Детальная информация
if data_month:
    print('📊 ЗАПРОС 6: Детальная информация по дням')
    print('-------------------------------------------')

    by_date = defaultdict(lambda: {'records': 0, 'skus': set(), 'expenses': []})
    for row in data_month:
        date = row['stat_date']
        by_date[date]['records'] += 1
        if row.get('sku'):
            by_date[date]['skus'].add(row['sku'])
        by_date[date]['expenses'].append(float(row.get('money_spent', 0)))

    print('')
    print('Дата       | Расход      | Записей | SKU | Все одинаковые?')
    print('-----------|-------------|---------|-----|----------------')
    for date in sorted(by_date.keys()):
        info = by_date[date]
        max_exp = max(info['expenses'])
        min_exp = min(info['expenses'])
        all_same = '✅' if max_exp == min_exp else '❌'
        print(f'{date} | {max_exp:>10.2f} ₽ | {info["records"]:>7} | {len(info["skus"]):>3} | {all_same}')

    print('')
    print(f'Итого дней с данными: {len(by_date)} из 31')
    print(f'Всего записей: {len(data_month)}')
    print('')

print('=====================================')
print('✅ Диагностика завершена!')
print('')
print('💡 Выводы:')
print('   Если "Всего дней с данными" < 31, значит проблема в синхронизации OZON')
print('   Если все 31 день есть, но сумма не совпадает - проблема в логике дедупликации')

File: run-diagnostics.mjs
#!/usr/bin/env node

/**
 * Скрипт для диагностики проблемы с расходами
 * Выполняет SQL запросы из debug-campaign-expenses.sql
 */

import { createClient } from '@supabase/supabase-js';

const SUPABASE_URL = "https://bkmicyguzlwampuindff.supabase.co";
const SUPABASE_ANON_KEY = "eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJzdXBhYmFzZSIsInJlZiI6ImJrbWljeWd1emx3YW1wdWluZGZmIiwicm9sZSI6ImFub24iLCJpYXQiOjE3NjQ2OTUwMjMsImV4cCI6MjA4MDI3MTAyM30.v8BlZ_k8DxdSmh5Ao1da7GHurSshE1cBsMxdfQCp9PQ";

const supabase = createClient(SUPABASE_URL, SUPABASE_ANON_KEY);

console.log('🔍 ДИАГНОСТИКА ПРОБЛЕМЫ С РАСХОДАМИ');
console.log('=====================================');
console.log('Кампания: Кабель КГ 2*2,5');
console.log('Период: 1-31 декабря 2025');
console.log('');

// Запрос 4: За неделю (1-7 декабря)
console.log('📊 ЗАПРОС 4: За неделю (1-7 декабря) - ожидается ~24 428 ₽');
console.log('----------------------------------------------------------------');
const { data: data4, error: error4 } = await supabase
  .from('ozon_performance_daily')
  .select('stat_date, money_spent')
  .eq('campaign_name', 'Кабель КГ 2*2,5')
  .gte('stat_date', '2025-12-01')
  .lte('stat_date', '2025-12-07');

if (error4) {
  console.error('❌ Ошибка:', error4.message);
} else {
  // Дедупликация по дням
  const dailyExpenses = new Map();
  data4?.forEach(row => {
    const date = row.stat_date;
    const expense = Number(row.money_spent || 0);
    dailyExpenses.set(date, Math.max(dailyExpenses.get(date) || 0, expense));
  });

  const total = Array.from(dailyExpenses.values()).reduce((sum, val) => sum + val, 0);
  console.log(`Всего дней: ${dailyExpenses.size}`);
  console.log(`Общая сумма: ${total.toFixed(2)} ₽`);
  console.log(`Ожидалось: ~24 428 ₽`);
  console.log(`Разница: ${(total - 24428).toFixed(2)} ₽`);
  console.log(`Совпадает: ${Math.abs(total - 24428) < 100 ? '✅ ДА' : '❌ НЕТ'}`);
}
console.log('');

// Запрос 5: За месяц (1-31 декабря)
console.log('📊 ЗАПРОС 5: За месяц (1-31 декабря) - ожидается ~109 130 ₽');
console.log('----------------------------------------------------------------');
const { data: data5, error: error5 } = await supabase
  .from('ozon_performance_daily')
  .select('stat_date, money_spent')
  .eq('campaign_name', 'Кабель КГ 2*2,5')
  .gte('stat_date', '2025-12-01')
  .lte('stat_date', '2025-12-31');

if (error5) {
  console.error('❌ Ошибка:', error5.message);
} else {
  // Дедупликация по дням
  const dailyExpenses = new Map();
  data5?.forEach(row => {
    const date = row.stat_date;
    const expense = Number(row.money_spent || 0);
    dailyExpenses.set(date, Math.max(dailyExpenses.get(date) || 0, expense));
  });

  const total = Array.from(dailyExpenses.values()).reduce((sum, val) => sum + val, 0);
  console.log(`Всего дней с данными: ${dailyExpenses.size} из 31`);
  console.log(`Общая сумма: ${total.toFixed(2)} ₽`);
  console.log(`Ожидалось: ~109 130 ₽`);
  console.log(`Разница: ${(total - 109130).toFixed(2)} ₽`);
  console.log(`Совпадает: ${Math.abs(total - 109130) < 100 ? '✅ ДА' : '❌ НЕТ'}`);
  console.log('');

  if (dailyExpenses.size < 31) {
    console.log('⚠️ ПРОБЛЕМА: Не все дни декабря есть в базе!');
    console.log(`   Отсутствует ${31 - dailyExpenses.size} дней`);
    console.log('');
    console.log('📅 Даты с данными:');
    const dates = Array.from(dailyExpenses.keys()).sort();
    dates.forEach((date, i) => {
      console.log(`   ${String(i + 1).padStart(2)}. ${date}: ${dailyExpenses.get(date).toFixed(2).padStart(10)} ₽`);
    });
    console.log('');

    // Найдем отсутствующие даты
    const allDates = [];
    for (let day = 1; day <= 31; day++) {
      allDates.push(`2025-12-${String(day).padStart(2, '0')}`);
    }
    const missingDates = allDates.filter(date => !dailyExpenses.has(date));
    if (missingDates.length > 0) {
      console.log('❌ Отсутствующие даты:');
      missingDates.forEach(date => console.log(`   - ${date}`));
      console.log('');
    }
  } else {
    console.log('✅ Все 31 день декабря присутствуют в базе');
  }
}
console.log('');

// Запрос 6: Детальная информация по дням
console.log('📊 ЗАПРОС 6: Детальная информация по дням');
console.log('-------------------------------------------');
const { data: data6, error: error6 } = await supabase
  .from('ozon_performance_daily')
  .select('stat_date, money_spent, sku, campaign_id')
  .eq('campaign_name', 'Кабель КГ 2*2,5')
  .gte('stat_date', '2025-12-01')
  .lte('stat_date', '2025-12-31')
  .order('stat_date', { ascending: true });

if (error6) {
  console.error('❌ Ошибка:', error6.message);
} else {
  const byDate = new Map();
  data6?.forEach(row => {
    const date = row.stat_date;
    if (!byDate.has(date)) {
      byDate.set(date, {
        records: 0,
        skus: new Set(),
        campaignIds: new Set(),
        expenses: []
      });
    }
    const info = byDate.get(date);
    info.records++;
    if (row.sku) info.skus.add(row.sku);
    if (row.campaign_id) info.campaignIds.add(row.campaign_id);
    info.expenses.push(Number(row.money_spent || 0));
  });

  console.log('');
  console.log('Дата       | Расход      | Записей | SKU | Campaign IDs | Все одинаковые?');
  console.log('-----------|-------------|---------|-----|--------------|----------------');
  Array.from(byDate.entries()).sort().forEach(([date, info]) => {
    const maxExpense = Math.max(...info.expenses);
    const minExpense = Math.min(...info.expenses);
    const allSame = maxExpense === minExpense ? '✅' : '❌';
    console.log(
      `${date} | ${maxExpense.toFixed(2).padStart(10)} ₽ | ${String(info.records).padStart(7)} | ${String(info.skus.size).padStart(3)} | ${String(info.campaignIds.size).padStart(12)} | ${allSame}`
    );
  });
  console.log('');
  console.log(`Итого дней с данными: ${byDate.size} из 31`);
  console.log(`Всего записей: ${data6?.length || 0}`);
}
console.log('');

// Запрос 7: Проверка orders_model и revenue_model
console.log('📊 ЗАПРОС 7: Проверка наличия model данных');
console.log('-------------------------------------------');
const { data: data7, error: error7 } = await supabase
  .from('ozon_performance_daily')
  .select('orders, orders_model, revenue, revenue_model')
  .eq('campaign_name', 'Кабель КГ 2*2,5')
  .gte('stat_date', '2025-12-01')
  .lte('stat_date', '2025-12-31');

if (error7) {
  console.error('❌ Ошибка:', error7.message);
} else {
  let hasOrdersModel = 0;
  let hasRevenueModel = 0;
  let totalOrders = 0;
  let totalOrdersModel = 0;
  let totalRevenue = 0;
  let totalRevenueModel = 0;

  data7?.forEach(row => {
    if (row.orders_model != null && row.orders_model > 0) hasOrdersModel++;
    if (row.revenue_model != null && row.revenue_model > 0) hasRevenueModel++;
    totalOrders += Number(row.orders || 0);
    totalOrdersModel += Number(row.orders_model || 0);
    totalRevenue += Number(row.revenue || 0);
    totalRevenueModel += Number(row.revenue_model || 0);
  });

  console.log(`Всего записей: ${data7?.length || 0}`);
  console.log(`Записей с orders_model > 0: ${hasOrdersModel}`);
  console.log(`Записей с revenue_model > 0: ${hasRevenueModel}`);
  console.log('');
  console.log(`Сумма orders: ${totalOrders}`);
  console.log(`Сумма orders_model: ${totalOrdersModel}`);
  console.log(`Сумма revenue: ${totalRevenue.toFixed(2)} ₽`);
  console.log(`Сумма revenue_model: ${totalRevenueModel.toFixed(2)} ₽`);
  console.log('');
  console.log(`ИТОГО заказов (orders + orders_model): ${totalOrders + totalOrdersModel}`);
  console.log(`ИТОГО выручка (revenue + revenue_model): ${(totalRevenue + totalRevenueModel).toFixed(2)} ₽`);
}
console.log('');

console.log('=====================================');
console.log('✅ Диагностика завершена!');
console.log('');
console.log('💡 Выводы:');
console.log('   Если "Всего дней с данными" < 31, значит проблема в синхронизации OZON');
console.log('   Если все 31 день есть, но сумма не совпадает - проблема в логике дедупликации');
